src/Features/Core/Portable/InheritanceMargin/Finders/BaseTypeSymbolsFinder.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.FindSymbols.FindReferences;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InheritanceMargin.Finders
+{
+    internal class BaseTypeSymbolsFinder : InheritanceSymbolsFinder
+    {
+        public static readonly BaseTypeSymbolsFinder Instance = new();
+
+        protected override Task<ImmutableArray<ISymbol>> GetAssociatedSymbolsAsync(
+            ISymbol symbol,
+            Solution solution,
+            CancellationToken cancellationToken)
+        {
+            var namedTypeSymbol = (INamedTypeSymbol)symbol;
+            if (namedTypeSymbol.BaseType == null)
+            {
+                // AllInterfaces are topologically sorted by default.
+                return Task.FromResult(namedTypeSymbol.AllInterfaces.CastArray<ISymbol>());
+            }
+
+            // Calculate indegree for all the symbols
+            using var _ = PooledDictionary<ISymbol, HashSet<ISymbol>>.GetInstance(out var indegreeSymbolsMapBuilder);
+            var baseTypes = BaseTypeFinder.FindBaseTypes(namedTypeSymbol);
+            for (var i = 0; i < baseTypes.Length; i++)
+            {
+                // baseTypes are order like,
+                //                        [0]          [1]           [2]
+                // symbol_we_search -> baseClass1 -> baseClass2 -> baseClass3 ...
+                // There is no edge points to the 'baseClass1'.
+                // And since interface can't have base class, the item before a baseClass is just the previous item in the array.
+                // e.g. 'baseClass2' could only be pointed by 'baseClass1'.
+                var baseType = baseTypes[i];
+                if (i == 0)
+                {
+                    indegreeSymbolsMapBuilder[baseType] = new HashSet<ISymbol>();
+                }
+                else
+                {
+                    indegreeSymbolsMapBuilder[baseType] = new HashSet<ISymbol>() { baseTypes[i - 1] };
```

Comment:
should these sets use MetaddataUnifying comparers?

Reply:
: ) Same comments above